pkg/compactor/blocks_cleaner.go

```diff
@@ -571,22 +571,30 @@ func listBlocksOutsideRetentionPeriod(idx *bucketindex.Index, threshold time.Tim
 	return
 }
 
-// findMostRecentModifiedTimeForBlock finds the most recent modification time for all files in a block.
-func findMostRecentModifiedTimeForBlock(ctx context.Context, blockID ulid.ULID, userBucket objstore.Bucket) (time.Time, error) {
-	var result time.Time
+var errStopIter = errors.New("stop iteration")
 
-	err := userBucket.Iter(ctx, blockID.String(), func(name string) error {
+// isPartialBlockOutsideDelayPeriod checks if the objects of a block were last modified before the provided cutoff time
+func isPartialBlockOutsideDelayPeriod(ctx context.Context, blockID ulid.ULID, userBucket objstore.InstrumentedBucket, partialDeletionCutoffTime time.Time) (bool, error) {
+	empty := true
```

Comment:
Can we still return and log most recent modified time, in case partial block is "outside" delay period.

Note that we only run this function on "partial blocks" -- it means there is always at least one object inside the block.

Reply:
> Note that we only run this function on "partial blocks" -- it means there is always at least one object inside the block.

I understand you're coming from the system view here of partial block -> soft delete -> hard delete. I'm accounting for the empty case as a matter of hygiene because soft deleting nothing doesn't have much of a point. I added the test case because the code accounts for this subtly with `!lastModified.IsZero()` and I initially flipped the case when refactoring.

When I change the function signature back for the log line the cases will be collapsed again, but I do think having the test case as a nudge for this is useful since it's easy to miss.

Reply:
Sounds good.